src/services/HmrcService.ts

```diff
@@ -74,32 +75,52 @@ export class HmrcService {
     	}
     }
 
-    async generateToken(): Promise<HmrcTokenResponse | undefined> {
+    async generateToken(backoffPeriodMs: number, maxRetries: number): Promise<HmrcTokenResponse | undefined> {
         
-    	try {
-    		const params = {
-    			client_secret : this.HMRC_CLIENT_SECRET,
-    			client_id : this.HMRC_CLIENT_ID,
-    			grant_type : "client_credentials",
-    		};
-    		const config: AxiosRequestConfig<any> = {
-    			headers: {
-    				Accept: "application/x-www-form-urlencoded",
-    			},
-    		};
-            
-    		const { data } = await axios.post(
-    			`${this.HMRC_BASE_URL}${Constants.HMRC_TOKEN_ENDPOINT_PATH}`,
-    			params,
-    			config,
-    		);
-
-    		this.logger.info("Received response from HMRC token endpoint");
-    		return data;
-    	} catch (error: any) {
-    		this.logger.error({ message: "An error occurred when generating HMRC token", hmrcErrorMessage: error.message, hmrcErrorCode: error.code, messageCode: MessageCodes.FAILED_GENERATING_HMRC_TOKEN });
-    		throw new AppError(HttpCodesEnum.SERVER_ERROR, "Error generating HMRC token");
+    	this.logger.debug("generateToken", HmrcService.name);
+
+    	let retryCount = 0;
+    	//retry for maxRetry count configured value if fails
+    	while (retryCount <= maxRetries) {
+    		this.logger.debug(`generateToken - trying to generate hmrcToken ${new Date().toISOString()}`, {
+    			retryCount,
+    		});
+    		try {
+    			const params = {
+    				client_secret : this.HMRC_CLIENT_SECRET,
+    				client_id : this.HMRC_CLIENT_ID,
+    				grant_type : "client_credentials",
+    			};
+    			const config: AxiosRequestConfig<any> = {
+    				headers: {
+    					"Content-Type": "application/x-www-form-urlencoded",
+    				},
+    			};
+				
+    			const { data } = await axios.post(
+    				`${this.HMRC_BASE_URL}${Constants.HMRC_TOKEN_ENDPOINT_PATH}`,
+    				params,
+    				config,
+    			);
+
+    			this.logger.info("Received response from HMRC token endpoint");
+    			return data;
+    		} catch (error: any) {
+    			this.logger.error({ message: "An error occurred when generating HMRC token", hmrcErrorMessage: error.message, hmrcStatusCode: error.status, hmrcErrorCode: error.code, messageCode: MessageCodes.FAILED_GENERATING_HMRC_TOKEN });				
+
+    			if (error.message.includes("500") && retryCount < maxRetries) {
```

Comment:
I'm surprised that it's the message that contains the 500 and not the response status code?